tools/fs/optimistic.js

```diff
@@ -174,19 +166,52 @@ const dependOnNodeModules = wrap(nodeModulesDir => {
   }
 });
 
+// Called when an optimistic function detects the given file does not
+// exist, but needs to return null or false rather than throwing an
+// exception. When/if the file is eventually created, we might only get a
+// file change notification for the parent directory, so it's important to
+// depend on the parent directory using this function, so that we don't
+// cache the unsuccessful result forever.
+function dependOnParentDirectory(path) {
+  const parentDir = pathDirname(path);
+  if (parentDir !== path) {
+    dependOnDirectory(parentDir);
+  }
+}
+
+// Called by any optimistic function that receives a */node_modules/* path
+// as its first argument, so that we can later bulk-invalidate the results
+// of those calls if the contents of the node_modules directory change.
+// Note that this strategy will not detect changes within subdirectories
+// of this node_modules directory, but that's ok because the use case we
+// care about is adding or removing npm packages.
+const dependOnNodeModules = wrap(nodeModulesDir => {
+  assert(pathIsAbsolute(nodeModulesDir));
+  assert(nodeModulesDir.endsWith(pathSep + "node_modules"));
+  return dependOnDirectory(nodeModulesDir);
+});
```

Comment:
This code already existed, but I saw an opportunity to reimplement it in terms of the new `dependOnDirectory` function.